### PR TITLE
Fast Formal Integral

### PR DESCRIPTION
**Pull request by Rodot-**
### :pencil: Speed up the GPU computation of the formal integral

**Type:**  :rocket: `feature`

This updates the current GPU implementation of the formal integral to parallelize over both frequency bin and impact parameter.

### :vertical_traffic_light: Testing

How did you test these changes?

- [x] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)

### :ballot_box_with_check: Checklist

- [x] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

### CODE REVIEW

File: tardis/transport/montecarlo/formal_integral_cuda.py [hunk 1]
```diff
@@ -12,6 +12,212 @@
 H_CGS = 6.62606957e-27
 
 
+@cuda.jit
+def cuda_vector_integrator(L, I_nu, N, R_max):
+
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/formal_integral_cuda.py:30]: docstring

File: tardis/transport/montecarlo/formal_integral_cuda.py [hunk 2]
```diff
@@ -12,6 +12,212 @@
...
+    ----------
+    r_inner : array(float64, 1d, C)
+        self.geometry.r_inner
+    r_outer : array(float64, 1d, C)
+        self.geometry.r_outer
+    time_explosion: float64
+        self.geometry.time_explosion
+    line_list_nu : array(float64, 1d, A)
+        self.plasma.line_list_nu
+    iT : np.float64
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/formal_integral_cuda.py]: These variables either need a proper description or a better name

Comment (andrewfullard) [reply tardis/transport/montecarlo/formal_integral_cuda.py]: (by "these variables" I mean everything from iT to z)

File: tardis/transport/montecarlo/formal_integral_cuda.py [hunk 3]
```diff
@@ -12,6 +12,212 @@
...
+    L : array(float64, 1d, C)
+        This is where the results will be stored
+    pp : array(float64, 1d, C)
+    exp_tau : array(float64, 1d, C)
+    I_nu array(floatt64, 2d, C)
+    z : array(float64, 2d, C)
+    shell_id : array(int64, 2d, C)
+    """
+
+    # todo: add all the original todos
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/formal_integral_cuda.py]: Probably out of date now...

File: tardis/transport/montecarlo/formal_integral_cuda.py [hunk 4]
```diff
@@ -12,6 +12,212 @@
...
+    pline = 0
+
+    nu = inu[nu_idx]
+
+    # now loop over discrete values along line
+
+    escat_contrib = 0.0
+    p = pp[p_idx]
+
+    # initialize z intersections for p values
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/formal_integral_cuda.py:140]: What are "z intersections"?

File: tardis/transport/montecarlo/formal_integral_cuda.py [hunk 5]
```diff
@@ -12,6 +12,212 @@
...
+    Jkkp = 0.0
+    pexp_tau = 0
+    patt_S_ul = 0
+    pJred_lu = 0
+    pJblue_lu = 0
+    pline = 0
+
+    nu = inu[nu_idx]
+
+    # now loop over discrete values along line
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/formal_integral_cuda.py:135]: Along which line? I don't see a loop here

File: tardis/transport/montecarlo/formal_integral_cuda.py [hunk 6]
```diff
@@ -12,6 +12,212 @@
...
+        )
+
+        for _ in range(max(nu_end_idx - pline, 0)):
+            # calculate e-scattering optical depth to next resonance point
+            zend = time_explosion / C_INV * (1.0 - line_list_nu[pline] / nu)
+            if first == 1:
+                # first contribution to integration
+                # NOTE: this treatment of I_nu_b (given
+                #   by boundary conditions) is not in Lucy 1999;
+                #   should be re-examined carefully
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/formal_integral_cuda.py:181]: Want to re-examine now? :D

File: tardis/transport/montecarlo/formal_integral_cuda.py [hunk 7]
```diff
@@ -236,52 +442,75 @@ def formal_integral(
...
             (inu_size, 2 * size_shell), dtype=np.int64
         )  # array(int64, 1d, C)
-        THREADS_PER_BLOCK = 32
-        blocks_per_grid = (inu_size // THREADS_PER_BLOCK) + 1
 
-        cuda_formal_integral[blocks_per_grid, THREADS_PER_BLOCK](
-            self.geometry.r_inner,
-            self.geometry.r_outer,
+        # These get separate names since they'll be copied back
+        d_L = cuda.device_array((inu_size,), dtype=np.float64)
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/formal_integral_cuda.py:277]: Separate but not very descriptive...

**Review by andrewfullard — CHANGES_REQUESTED** [review]
Mostly documentation-related requests.

File: tardis/transport/montecarlo/formal_integral_cuda.py (hunk 4 above)
Comment (Rodot-) [reply tardis/transport/montecarlo/formal_integral_cuda.py:140]: https://tardis-sn.github.io/tardis/physics/spectrum/formal_integral.html#calculating-i-nu-p

File: tardis/transport/montecarlo/formal_integral_cuda.py (hunk 5 above)
Comment (Rodot-) [reply tardis/transport/montecarlo/formal_integral_cuda.py:135]: Yeah, the loop is in the GPU threads, but I'll remove the comment

File: tardis/transport/montecarlo/formal_integral_cuda.py (hunk 3 above)
Comment (Rodot-) [reply tardis/transport/montecarlo/formal_integral_cuda.py]: Yeah, I have no idea even where they are anymore.  I'll remove it

File: tardis/transport/montecarlo/formal_integral_cuda.py (hunk 7 above)
Comment (Rodot-) [reply tardis/transport/montecarlo/formal_integral_cuda.py:277]: Generally, it is standard to prepend device objects with a 'd'

File: tardis/transport/montecarlo/formal_integral_cuda.py (hunk 2 above)
Comment (Rodot-) [reply tardis/transport/montecarlo/formal_integral_cuda.py]: I'll work on giving them better descriptions, will have to pin some of them down

File: tardis/transport/montecarlo/formal_integral_cuda.py (hunk 7 above)
Comment (Rodot-) [reply tardis/transport/montecarlo/formal_integral_cuda.py:270]: I'll make a note

File: tardis/transport/montecarlo/formal_integral_cuda.py (hunk 2 above)
Comment (Rodot-) [reply tardis/transport/montecarlo/formal_integral_cuda.py:72]: I've updated the descriptions, let me know what you think

**Review by andrewfullard — APPROVED** [review]

File: tardis/spectrum/formal_integral_cuda.py [hunk 8]
```diff
@@ -41,44 +62,59 @@ def cuda_formal_integral(
...
     Jblue_lu : array(float64, 1d, C)
+        J estimator from blue end of the line from lower to upper level
     tau_sobolev : array(float64, 2d, C)
+        Sobolev Optical depth for each line in each shell
     electron_density : array(float64, 1d, C)
+        electron density in each shell
     N : int64
+        Number of impact parameter values (p)
     L : array(float64, 1d, C)
+        Luminosity density at each frequencyt
```
Comment (DeerWhale) [inline tardis/spectrum/formal_integral_cuda.py:91]:
```suggestion
        Luminosity density at each frequency
```

**Review by andrewfullard — DISMISSED** [review]

**Review by andrewfullard — APPROVED** [review]

**Pull request merged**